lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2508,6 +2516,116 @@ signTransaction tl preferredLatestEra keyLookup (rootKey, rootPwd) utxo =
             keyLookup
             inputResolver
 
+type MakeRewardAccountBuilder k =
+    (k 'RootK XPrv, Passphrase "encryption") -> (XPrv, Passphrase "encryption")
+
+-- | Produce witnesses and construct a transaction from a given selection.
+--
+-- Requires the encryption passphrase in order to decrypt the root private key.
+-- Note that this doesn't broadcast the transaction to the network. In order to
+-- do so, use 'submitTx'.
+--
+buildAndSignTransactionNoSelection
+    :: forall (n :: NetworkDiscriminant)
+     . (Typeable n, DelegationAddress n ShelleyKey 'CredFromKeyK)
+    => Tracer IO WalletLog
+    -> TimeInterpreter (Either PastHorizonException)
+    -> DBLayer IO (SeqState n ShelleyKey) ShelleyKey
+    -> NetworkLayer IO Block
+    -> TransactionLayer ShelleyKey 'CredFromKeyK SealedTx
+    -> WalletId
+    -> AnyRecentEra
+    -> Passphrase "user"
+    -> TransactionCtx
+    -> IO (Tx, TxMeta, SealedTx)
+buildAndSignTransactionNoSelection
+    tr ti db@DBLayer{..} netLayer txLayer
+    walletId era pwd txContext = WriteTx.withRecentEra era $
+    \(_recentEra :: WriteTx.RecentEra recentEra) -> do
+        utxo@(_utxoIndex, wallet, _history) <- atomically $ do
+            checkpoint <- readCheckpoint walletId >>=
+                maybe (throwIO
+                    $ ExceptionConstructTx
+                    $ ErrConstructTxNoSuchWallet
+                    $ ErrNoSuchWallet walletId) pure
+            pending  <-
+                readTxHistory
+                    walletId Nothing Descending wholeRange (Just Pending)
+            let pendingTxs = Set.fromList (fromTransactionInfo <$> pending)
+            pure
+                ( UTxOIndex.fromMap
+                    $ CS.toInternalUTxOMap
+                    $ availableUTxO pendingTxs checkpoint
+                , checkpoint
+                , pendingTxs
+                )
+
+        protocolParams <- currentProtocolParameters netLayer
+        unsignedTxBody <-
+            either (throwIO . (ExceptionConstructTx . ErrConstructTxBody)) pure $
+                mkUnsignedTransaction txLayer @recentEra
+                    (getRawKey (Seq.rewardAccountKey (getState wallet)))
+                    protocolParams txContext (Left (PreSelection []))
+
+        unsignedBalancedTx <- balanceTx protocolParams utxo unsignedTxBody
+        signedSealedTx <- signBalancedTx wallet unsignedBalancedTx
+
+        let ( tx
+                , _tokenMapWithScripts1
+                , _tokenMapWithScripts2
+                , _certificates
+                , _validityIntervalExplicit
+                , _witnessCount
+                ) = decodeTx txLayer anyCardanoEra signedSealedTx
+        let amountIn = F.fold $ snd <$> tx ^. #resolvedInputs
```

Comment:
Unfortunately, the `resolved` part of `resolvedInputs` is from jormungandr times and inconsistently constructed. From `decodeTx` it is set to [`Coin 0`](https://github.com/input-output-hk/cardano-wallet/blob/288094b8c67a70dd6d491bbc334d71175f4c3981/lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Babbage.hs#L110). 

I wonder if we can't use [`applyTxToUTxO`](https://github.com/input-output-hk/cardano-wallet/blob/288094b8c67a70dd6d491bbc334d71175f4c3981/lib/wallet/src/Cardano/Wallet/Primitive/Model.hs#L520-L524) though 🤔 We'd compare the UTxO balance before/after. This should be ideal from a code re-use and consistency point of view. Since rewards aren't UTxOs we probably still need some custom logic for it here though (like in `buildAndSignTransaction`).

Reply:
I believe this is the cause of the integration test failure:

<details>
<summary>STAKE_POOLS_JOIN_01rewards: expected: Outgoing but got: Incoming</summary>

```haskell
  integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:226:52:
  2) API Specifications, SHELLEY_STAKE_POOLS, STAKE_POOLS_JOIN_01rewards - Can join a pool, earn rewards and collect them
       While verifying value:
         ( Status
             { statusCode = 202
             , statusMessage = "Accepted"
             }
         , Right
             ( ApiTransaction
                 { id = ApiT
                     ( Hash "\x1b\x959`r8cèíªDk3\x15²á:\x97è¢\x1b¯&>\x1f?VMlW}á" )
                 , amount = Quantity 99998847000
                 , fee = Quantity 153000
                 , depositTaken = Quantity 0
                 , depositReturned = Quantity 0
                 , insertedAt = Nothing
                 , pendingSince = Just
                     ( ApiBlockReference
                         { absoluteSlotNumber = ApiT
                             ( SlotNo 54 )
                         , slotId = ApiSlotId
                             { epochNumber = ApiT
                                 ( EpochNo
                                     { unEpochNo = 0 }
                                 )
                             , slotNumber = ApiT
                                 ( SlotInEpoch
                                     { unSlotInEpoch = 54 }
                                 )
                             }
                         , time = 2022-12-22 13:31:38.8 UTC
                         , block = ApiBlockInfo
                             ( Quantity 15 )
                         }
                     )
                 , expiresAt = Just
                     ( ApiSlotReference
                         { absoluteSlotNumber = ApiT
                             ( SlotNo 36056 )
                         , slotId = ApiSlotId
                             { epochNumber = ApiT
                                 ( EpochNo
                                     { unEpochNo = 225 }
                                 )
                             , slotNumber = ApiT
                                 ( SlotInEpoch
                                     { unSlotInEpoch = 56 }
                                 )
                             }
                         , time = 2022-12-22 15:31:39.2 UTC
                         }
                     )
                 , depth = Nothing
                 , direction = ApiT Incoming
                 , inputs = []
                 , outputs =
                     [ AddressAmount
                         { address =
                             ( ApiT
                                 ( Address "\x1Môö\x92\x16a\x13ù5%[\x8eL,f}®PP;àOîôJØ\x11õò»\x9c\x91"Ù\x91.
                                 2Q\x15êq.\x8aÝ*Á\x96,_@Ë&\xcò\x80" )
                             , Proxy
                             )
                         , amount = Quantity 99998847000
                         , assets = ApiT
                             ( TokenMap
                                 ( fromList [] )
                             )
                         }
                     ]
                 , collateral = []
                 , collateralOutputs = ApiAsArray Nothing
                 , withdrawals = []
                 , status = ApiT Pending
                 , metadata = Nothing
                 , scriptValidity = Just ( ApiT TxScriptValid )
                 , certificates =
                     [ WalletDelegationCertificate
                         ( RegisterRewardAccount
                             { rewardAccountPath = ApiT
                                 ( DerivationIndex 2147485500 ) :|
                                 [ ApiT
                                     ( DerivationIndex 2147485463 )
                                 , ApiT
                                     ( DerivationIndex 2147483648 )
                                 , ApiT
                                     ( DerivationIndex 2 )
                                 , ApiT
                                     ( DerivationIndex 0 )
                                 ]
                             }
                         )
                     , WalletDelegationCertificate
                         ( JoinPool
                             { rewardAccountPath = ApiT
                                 ( DerivationIndex 2147485500 ) :|
                                 [ ApiT
                                     ( DerivationIndex 2147485463 )
                                 , ApiT
                                     ( DerivationIndex 2147483648 )
                                 , ApiT
                                     ( DerivationIndex 2 )
                                 , ApiT
                                     ( DerivationIndex 0 )
                                 ]
                             , pool = ApiT
                                 ( PoolId
                                     { getPoolId = "´WhÁ¢ÚKÑ>¼ª\x1e¥\x14\x8í£\x1dÌ!v\ËÔ\x7Í©ò" }
                                 )
                             }
                         )
                     ]
                 , mint = ApiAssetMintBurn
                     { tokens = []
                     , walletPolicyKeyHash = Nothing
                     , walletPolicyKeyIndex = Nothing
                     }
                 , burn = ApiAssetMintBurn
                     { tokens = []
                     , walletPolicyKeyHash = Nothing
                     , walletPolicyKeyIndex = Nothing
                     }
                 , validityInterval = Just
                     ( ApiValidityIntervalExplicit
                         ( ValidityIntervalExplicit
                             { invalidBefore = Quantity 0
                             , invalidHereafter = Quantity 36056
                             }
                         )
                     )
                 , scriptIntegrity = Nothing
                 , extraSignatures = []
                 }
             )
         )
       expected: Outgoing
        but got: Incoming
```

</details>

(presumably from `amountIn == Coin 0`)

which I got when trying to actually use `buildAndSignTransactionNoSelection`.

Reply:
> I wonder if we can't use [applyTxToUTxO](https://github.com/input-output-hk/cardano-wallet/blob/288094b8c67a70dd6d491bbc334d71175f4c3981/lib/wallet/src/Cardano/Wallet/Primitive/Model.hs#L520-L524) though 🤔

No, we need to have some concept of "ours" / filtering, so it would probably be: [`applyOurTxToUTxO `](https://github.com/input-output-hk/cardano-wallet/blob/288094b8c67a70dd6d491bbc334d71175f4c3981/lib/wallet/src/Cardano/Wallet/Primitive/Model.hs#L715-L731) in that case